src/compiler/scala/tools/nsc/typechecker/Namers.scala

```diff
@@ -710,6 +710,9 @@ trait Namers extends MethodSynthesis {
       if (isScala) {
         if (nme.isSetterName(tree.name)) ValOrVarWithSetterSuffixError(tree)
         if (tree.mods.isPrivateLocal && tree.mods.isCaseAccessor) PrivateThisCaseClassParameterError(tree)
+        if (tree.mods.isImplicit && !owner.isTerm && settings.warnImplicitNoAnnotation && tree.tpt.isEmpty) {
+          reporter.warning(tree.pos, s"implicit value ${tree.name} has no type annotation")
```

Comment:
The term "type annotation" is widely understood but the spec talks about "result types" in this context (and about "type ascriptions" after expressions).

A precedent for wording of this sort of message is:

```
scala> def foo = { return 1; 1 }
<console>:11: error: method foo has return statement; needs result type
       def foo = { return 1; 1 }
                   ^
```

`s/needs/should have a` and I think it fits here.